src/node_api.cc

```diff
@@ -18,16 +18,32 @@ static
 napi_status napi_clear_last_error(napi_env env);
 
 struct napi_env__ {
-  explicit napi_env__(v8::Isolate* _isolate, uv_loop_t* _loop)
-      : isolate(_isolate),
-        last_error(),
-        loop(_loop) {}
-  v8::Isolate* isolate;
+  explicit napi_env__(v8::Local<v8::Context> context)
+      : isolate(context->GetIsolate()),
+        context_persistent(isolate, context) {
+    CHECK_EQ(isolate, context->GetIsolate());
+    CHECK_NOT_NULL(node_env());
+  }
+
+  v8::Isolate* isolate;  // Shortcut for context()->GetIsolate()
+  node::Persistent<v8::Context> context_persistent;
+
+  inline v8::Local<v8::Context> context() const {
+    return StrongPersistentToLocal(context_persistent);
+  }
+
+  inline node::Environment* node_env() const {
+    return node::Environment::GetCurrent(context());
```

Comment:
Are there cases where this can change? If not, L28 could be const, and this could be cached as well.

Reply:
> Are there cases where this can change?

Not currently, no.

> If not, L28 could be const

Yup, done!

> and this could be cached as well.

The lookup isn’t so expensive + frequent that we would need to do that, I think.

We could cache it on the `ThreadSafeFunction` function if you feel strongly about it